Userland/Libraries/LibSoftGPU/Device.cpp

```diff
@@ -683,19 +683,30 @@ void Device::draw_primitives(PrimitiveType primitive_type, FloatMatrix4x4 const&
 
                     FloatVector4 vertex_to_light;
 
+                    if ((vertex.eye_coordinates.w() != 0.0f) && (light.position.w() == 0.0f)) {
+                        vertex_to_light = light.position;
+                    } else if ((vertex.eye_coordinates.w() == 0.0f) && (light.position.w() != 0.0f)) {
+                        vertex_to_light = -vertex.eye_coordinates;
+                    } else {
+                        // FIXME: The OpenGL 1.5 spec says nothing about the case where P1 and P2 BOTH have a w value of 1, which would
+                        // then mean the light position has an implicit value of (0, 0, 0, 0). This doesn't make any logical sense, and it most likely
+                        // a typographical error. Most other GL implementations seem to just fix it to the distance from the vertex to the light, which
+                        // seems to work just fine.
+                        // If somebody with more insight about this could clarify this eventually, that'd be great.
+                        vertex_to_light = (light.position - vertex.eye_coordinates);
+                    }
+
```

Comment:
I think it would be great to put this functionality into a lambda expression. This is used in a few places in the GL spec for lighting.
This should then take 2 Vector4s and return a Vector3.

Reply:
Similarly, the dot product operator (Circle with a dot) is defined to return `max(0, dot(a, b))` Maybe extract that as well to stay consistent.